pkg/database/truncator.go

```diff
@@ -0,0 +1,171 @@
+)
+
+// Truncator provides truncation against an underlying storage
+// of appendable data.
+type Truncator interface {
+	// Plan returns the transaction upto which the log can be truncated.
+	// When resulting transaction before specified time does not exists
+	//  * No transaction header is returned.
+	//  * Returns nil TxHeader, and an error.
+	Plan(time.Time) (*store.TxHeader, error)
+
+	// Truncate runs truncation against the relevant appendable logs. Must
+	// be called after result of Plan().
+	Truncate(*store.TxHeader) error
+}
+
+func NewVlogTruncator(d DB) Truncator {
+	return &vlogTruncator{
+		db:      d.(*db),
+		metrics: newTruncatorMetrics(d.GetName()),
+	}
+}
+
+// vlogTruncator implements Truncator for the value-log appendable
+type vlogTruncator struct {
+	db      *db
+	metrics *truncatorMetrics
+}
+
+// Plan returns the transaction upto which the value log can be truncated.
+// When resulting transaction before specified time does not exists
+//  * No transaction header is returned.
+//  * Returns nil TxHeader, and an error.
+// ts time is truncated to the start of the day.
+func (v *vlogTruncator) Plan(ts time.Time) (*store.TxHeader, error) {
+	hdr, err := v.db.st.FirstTxSince(ts)
+	if err != nil {
+		return nil, err
+	}
+
+	// if the transaction is on or before the retention period, then we can truncate upto this
+	// transaction otherwise, we cannot truncate since the retention period has not been reached
+	// and truncation would otherwise add an extra transaction to the log for sql catalogue.
+	err = v.isRetentionPeriodReached(ts, time.Unix(hdr.Ts, 0))
+	if err != nil {
+		return nil, err
+	}
+	return hdr, nil
+}
+
+// isRetentionPeriodReached returns an error if the retention period has not been reached.
+func (v *vlogTruncator) isRetentionPeriodReached(retentionPeriod time.Time, txTimestamp time.Time) error {
+	txTime := TruncateToDay(txTimestamp)
+	if txTime.Unix() <= retentionPeriod.Unix() {
+		return nil
+	}
+	return ErrRetentionPeriodNotReached
+}
+
+// commitCatalog commits the current sql catalogue as a new transaction.
+func (v *vlogTruncator) commitCatalog(hdr *store.TxHeader) (*store.TxHeader, error) {
+	// copy sql catalogue
+	ctx := context.Background()
+	tx, err := v.db.st.NewTx(ctx, store.DefaultTxOptions())
+	if err != nil {
+		return nil, err
+	}
+
+	err = v.db.CopyCatalogToTx(ctx, tx)
+	if err != nil {
+		v.db.Logger.Errorf("error during truncation for database '%s' {err = %v, id = %v, type=sql_catalogue_copy}", v.db.name, err, hdr.ID)
+		return nil, err
+	}
+	defer tx.Cancel()
+
+	// setting the metadata to record the transaction upto which the log was truncated
+	tx.WithMetadata(store.NewTxMetadata().WithTruncatedTxID(hdr.ID))
+
+	// commit catalogue as a new transaction
+	return tx.Commit(context.Background())
+}
+
+// Truncate runs truncation against the relevant appendable logs upto the specified transaction offset.
+func (v *vlogTruncator) Truncate(hdr *store.TxHeader) error {
```

Comment:
minor, TruncateUptoTx naming as in the embedded store is more descriptive

Reply:
header struct doesn't seem to be needed in this method but just the txID

Reply:
fixed